### PR TITLE
Add support for a default and default_factory with strawberry.field

### LINKED ISSUE #569: Parity with `dataclasses.field`
While starting to use Strawberry 🍓  and admittedly being a beginner, I noticed I had to make a choice : I have to choose between Strawberry features on my fields via `strawberry.field` (e.g. adding a description) and using native `dataclasses.field` features.

Some of the native `dataclasses.field` attributes that I do use on my dataclasses are `default_factory` and `metadata` (which allows me to pass an arbitrary dictionary of extra settings to my field)

Is there a reason why `strawberry.field` doesn't directly extend `dataclasses.field` ? Wouldn't that be more pythonic ? 😺 

Practically speaking:
```python
import faker

# https://faker.readthedocs.io/en/master/
fake = faker.Faker()

@strawberry.type
class User:
    first_name: str = strawberry.field(description="First name",  default_factory=fake.first_name, metadata={"foo": "bar"})
```

This isn't currently possible because `@strawberry.field` doesn't accept `default_factory` and `metadata`.
Would've been cool though, especially in the context of being able to generate fake data on the fly

### PR DESCRIPTION
**Pull request by jkimbo**
## Description

Add `default` and `default_factory` arguments to `strawberry.field` to bring it inline with `dataclasses.field`.

Note: I haven't added the `metadata` attribute because I'm not sure there is a compelling reason to add it.

# Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Closes #569

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Add support for `default` and `default_factory` arguments in `strawberry.field`

```python
@strawberry.type
class Droid:
    name: str = strawberry.field(default="R2D2")
    aka: List[str] = strawberry.field(default_factory=["Artoo"])
```

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Add support for `default` and `default_factory` arguments in `strawberry.field`

```python
@strawberry.type
class Droid:
    name: str = strawberry.field(default="R2D2")
    aka: List[str] = strawberry.field(default_factory=["Artoo"])
```

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Add support for `default` and `default_factory` arguments in `strawberry.field`

```python
@strawberry.type
class Droid:
    name: str = strawberry.field(default="R2D2")
    aka: List[str] = strawberry.field(default_factory=["Artoo"])
```

File: RELEASE.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Release type: minor
+
+Add support for `default` and `default_factory` arguments in `strawberry.field`
+
+```python
+@strawberry.type
+class Droid:
+    name: str = strawberry.field(default="R2D2")
+    aka: List[str] = strawberry.field(default_factory=["Artoo"])
```
Comment (patrick91) [inline RELEASE.md:9]: should this be? 
```suggestion
    aka: List[str] = strawberry.field(default_factory=lambda: ["Artoo"])
```

File: strawberry/exceptions.py [hunk 2]
```diff
@@ -143,3 +143,23 @@ def __init__(self, expected: int, received: int):
...
+
+
+class FieldWithResolverAndDefaultFactoryError(Exception):
+    def __init__(self, field_name: str, type_name: str):
+        message = (
+            f'Field "{field_name}" on type "{type_name}" cannot define a '
+            "default_factory and a resolver."
+        )
+
+        super().__init__(message)
```
Comment (patrick91) [inline strawberry/exceptions.py:165]: Nice!

File: strawberry/field.py [hunk 3]
```diff
@@ -39,6 +39,7 @@ def __init__(
         base_resolver: Optional[StrawberryResolver] = None,
         permission_classes: List[Type[BasePermission]] = (),  # type: ignore
         default_value: Any = undefined,
+        default_factory: Union[Callable, object] = undefined,
```
Comment (patrick91) [inline strawberry/field.py:42]: can we use `UNSET` instead of undefined? 😊

**Review by patrick91 — COMMENTED** [review]
Looks good! I was going to suggest checking how the defaults values were printed in the schema, but they are only printed for arguments (and we support that already). So all good from my side once resolved those two minor comments 😊

File: RELEASE.md (hunk 1 above)
Comment (jkimbo) [reply RELEASE.md:9]: Yes it should be!

File: strawberry/field.py (hunk 3 above)
Comment (jkimbo) [reply strawberry/field.py:42]: Sure. Should I change `default_value` as well?

Comment (patrick91) [reply strawberry/field.py:42]: yes, then we can remove this: https://github.com/strawberry-graphql/strawberry/blob/44dd42d4f442eb77078c8def079b22ed4800c583/strawberry/types/type_resolver.py#L31-L32

but if it starts looking like a big change we can defer it to another PR :)

**Review by patrick91 — APPROVED** [review]
💯

**Comment by patrick91** [comment]
@jkimbo will fix the pre-commit issue (just send a PR to update our config too)

**Pull request merged**